### PR TITLE
HPF-34:Support system that doesn't have ULLONG_MAX.

### PR DESCRIPTION
**Pull request by hyoklee**
For example, Jam/Koala needs this patch.

### CODE REVIEW

**Review by derobins — CHANGES_REQUESTED** [review]
Like the other PRs for missing constants, ULLONG_MAX is c99, which is required for building HDF5. gcc 4.4 handles c99 just fine and the problem is usually the C++ wrappers, as g++ 4.4 needs special handling to bring in the C99 constants.

In the next week or so, I'm going to add -std=c++11 to the C++ wrappers so we can start bringing in modern C++ language features and then jam/koala are *definitely* going to have severe problems with the C++ wrappers.

File: src/H5public.h [hunk 1]
```diff
@@ -254,6 +254,10 @@ typedef long long ssize_t;
 /* int64_t type is used for creation order field for links.  It may be
  * defined in Posix.1g, otherwise it is defined here.
  */
+#ifndef ULLONG_MAX
+#define ULLONG_MAX ((haddr_t)(long long)(-1))
+#endif
```
Comment (derobins) [inline src/H5public.h:259]: Also, fwiw, having ULLONG_MAX be cast to a haddr_t is awkward.

**Comment by hyoklee** [comment]
> Like the other PRs for missing constants, ULLONG_MAX is c99, which is required for building HDF5. gcc 4.4 handles c99 just fine and the problem is usually the C++ wrappers, as g++ 4.4 needs special handling to bring in the C99 constants.
> 
> In the next week or so, I'm going to add -std=c++11 to the C++ wrappers so we can start bringing in modern C++ language features and then jam/koala are _definitely_ going to have severe problems with the C++ wrappers.

Update FAQ / user guide /  configure script first so users will know that HDF5 won't be compiled on the old versions of compilers before writing any further code.

**Comment by derobins** [comment]
> > Like the other PRs for missing constants, ULLONG_MAX is c99, which is required for building HDF5. gcc 4.4 handles c99 just fine and the problem is usually the C++ wrappers, as g++ 4.4 needs special handling to bring in the C99 constants.
> > In the next week or so, I'm going to add -std=c++11 to the C++ wrappers so we can start bringing in modern C++ language features and then jam/koala are _definitely_ going to have severe problems with the C++ wrappers.
> 
> Update FAQ / user guide / configure script first so users will know that HDF5 won't be compiled on the old versions of compilers before writing any further code.

From RELEASE.txt (develop flavor). gcc 4.1.2 < 4.4.7 (honestly, even 4.4.7 is a stretch and we should drop that).

Supported Platforms                                                              
===================                                                              
                                                                                 
    Linux 2.6.32-696.16.1.el6.ppc64 gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)  
    #1 SMP ppc64 GNU/Linux        g++ (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)    
    (ostrich)                     GNU Fortran (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)
                                  IBM XL C/C++ V13.1                             
                                  IBM XL Fortran V15.1                           
                                                                                 
    Linux 3.10.0-327.10.1.el7   GNU C (gcc), Fortran (gfortran), C++ (g++)       
    #1 SMP x86_64 GNU/Linux       compilers:                                     
    (kituo/moohan)                Version 4.8.5 20150623 (Red Hat 4.8.5-4)       
                                    Version 4.9.3, Version 5.2.0,                
                                  Intel(R) C (icc), C++ (icpc), Fortran (icc)    
                                  compilers:                                     
                                     Version 17.0.0.098 Build 20160721           
                                  MPICH 3.1.4 compiled with GCC 4.9.3            
                                                                                 
    SunOS 5.11 32- and 64-bit     Sun C 5.12 SunOS_sparc                         
    (emu)                         Sun Fortran 95 8.6 SunOS_sparc                 
                                  Sun C++ 5.12 SunOS_sparc                       
                                                                                 
    Windows 10 x64                Visual Studio 2015 w/ Intel Fortran 18 (cmake) 
                                  Visual Studio 2017 w/ Intel Fortran 19 (cmake) 
                                  Visual Studio 2019 w/ Intel Fortran 19 (cmake) 
                                  Visual Studio 2019 w/ MSMPI 10.1 (cmake)       
                                                                                 
    Mac OS X Yosemite 10.10.5     Apple clang/clang++ version 6.1 from Xcode 7.0 
    64-bit                        gfortran GNU Fortran (GCC) 4.9.2               
    (osx1010dev/osx1010test)      Intel icc/icpc/ifort version 15.0.3            
                                                                                 
    Mac OS X El Capitan 10.11.6   Apple clang/clang++ version 7.3.0 from Xcode 7.3
    64-bit                        gfortran GNU Fortran (GCC) 5.2.0               
    (osx1011dev/osx1011test)      Intel icc/icpc/ifort version 16.0.2            
                                                                                 
    Mac OS Sierra 10.12.6         Apple LLVM version 8.1.0 (clang/clang++-802.0.42)
    64-bit                        gfortran GNU Fortran (GCC) 7.1.0               
    (swallow/kite)                Intel icc/icpc/ifort version 17.0.2

**Comment by hyoklee** [comment]
> > > Like the other PRs for missing constants, ULLONG_MAX is c99, which is required for building HDF5. gcc 4.4 handles c99 just fine and the problem is usually the C++ wrappers, as g++ 4.4 needs special handling to bring in the C99 constants.
> > > In the next week or so, I'm going to add -std=c++11 to the C++ wrappers so we can start bringing in modern C++ language features and then jam/koala are _definitely_ going to have severe problems with the C++ wrappers.
> > 
> > 
> > Update FAQ / user guide / configure script first so users will know that HDF5 won't be compiled on the old versions of compilers before writing any further code.
> 
> From RELEASE.txt (develop flavor). gcc 4.1.2 < 4.4.7 (honestly, even 4.4.7 is a stretch and we should drop that).
> 
> # Supported Platforms
> ```
> Linux 2.6.32-696.16.1.el6.ppc64 gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)  
> #1 SMP ppc64 GNU/Linux        g++ (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)    
> (ostrich)                     GNU Fortran (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18)
>                               IBM XL C/C++ V13.1                             
>                               IBM XL Fortran V15.1                           
>                                                                              
> Linux 3.10.0-327.10.1.el7   GNU C (gcc), Fortran (gfortran), C++ (g++)       
> #1 SMP x86_64 GNU/Linux       compilers:                                     
> (kituo/moohan)                Version 4.8.5 20150623 (Red Hat 4.8.5-4)       
>                                 Version 4.9.3, Version 5.2.0,                
>                               Intel(R) C (icc), C++ (icpc), Fortran (icc)    
>                               compilers:                                     
>                                  Version 17.0.0.098 Build 20160721           
>                               MPICH 3.1.4 compiled with GCC 4.9.3            
>                                                                              
> SunOS 5.11 32- and 64-bit     Sun C 5.12 SunOS_sparc                         
> (emu)                         Sun Fortran 95 8.6 SunOS_sparc                 
>                               Sun C++ 5.12 SunOS_sparc                       
>                                                                              
> Windows 10 x64                Visual Studio 2015 w/ Intel Fortran 18 (cmake) 
>                               Visual Studio 2017 w/ Intel Fortran 19 (cmake) 
>                               Visual Studio 2019 w/ Intel Fortran 19 (cmake) 
>                               Visual Studio 2019 w/ MSMPI 10.1 (cmake)       
>                                                                              
> Mac OS X Yosemite 10.10.5     Apple clang/clang++ version 6.1 from Xcode 7.0 
> 64-bit                        gfortran GNU Fortran (GCC) 4.9.2               
> (osx1010dev/osx1010test)      Intel icc/icpc/ifort version 15.0.3            
>                                                                              
> Mac OS X El Capitan 10.11.6   Apple clang/clang++ version 7.3.0 from Xcode 7.3
> 64-bit                        gfortran GNU Fortran (GCC) 5.2.0               
> (osx1011dev/osx1011test)      Intel icc/icpc/ifort version 16.0.2            
>                                                                              
> Mac OS Sierra 10.12.6         Apple LLVM version 8.1.0 (clang/clang++-802.0.42)
> 64-bit                        gfortran GNU Fortran (GCC) 7.1.0               
> (swallow/kite)                Intel icc/icpc/ifort version 17.0.2   
> ```

This is not enough. Configure script should stop as soon as it detects an old compiler.

**Review by qkoziol — CHANGES_REQUESTED** [review]
ULLONG_MAX is in the C99 standard, it shouldn't be defined if the compiler doesn't support it.   Downstream header files (included after a "#include <hdf5.h>" in user source code) would be misled by HDF5 defining it.

**Pull request closed without merging**